keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,414 @@
+* Implement in-cluster logic to handle node/control plane partitioning. 
+* Enable detach for all the storage providers.
+* Existing in-tree volumes are not targeted by this proposal.
+
+## Proposal
+
+User stories:
+
+* As a cluster administrator I want my stateful workload to failover in case a node is partitioned from the control plane without any intervention when it's safe to do so.
+
+* As a developer I would like to rely on a self-healing platform for my stateful workloads.
+
+This KEP plans to introduce the following CSI capabilities as described in this [PR](https://github.com/container-storage-interface/spec/pull/477).
+
+* In the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md), introduce `UNPUBLISH_FENCE` controller service capability and `FORCE_UNPUBLISH` node service capability.
+  * The `UNPUBLISH_FENCE` controller service capability indicates that the SP supports ControllerUnpublishVolume.fence field.
+  * The `FORCE_UNPUBLISH` node service capability indicates that the SP supports the NodeUnpublishVolume.force field. It also indicates that the SP supports the NodeUnstageVolume.force field if it also has the STAGE_UNSTAGE_VOLUME node service capability.
+
+* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+  // SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations. When set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking whether a volume
+  // can be detached safely
+  // +optional
+  SafeDetach *bool
+}
+
+```
+
+Existing logic:
+* When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+* After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+* The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.
+
+* Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:
+
+  * pods that are backed by a CSI driver that supports `SafeDetach`.
+  * pods that are backed by volumes that aren't marked as `ReadWriteMany`. This would avoid breaking having multiple replicas of the same pod.
```

Comment:
"breaking having"? what is this trying to say?

Reply:
That's a typo.  Will fix it.

Reply:
Fixed.